msal/broker.py

```diff
@@ -0,0 +1,198 @@
+            # https://docs.microsoft.com/en-us/azure/active-directory/devices/troubleshoot-device-dsregcmd
+        "error_description": "{}. Status: {}, Error code: {}, Tag: {}".format(
+            context,
+            error.get_status(), error.get_error_code(), error.get_tag()),
+        }
+
+
+def _read_account_by_id(account_id):
+    """Return the callback result which contains the account or error"""
+    callback_data = _CallbackData()
+    pymsalruntime.read_account_by_id(
+        account_id,
+        "correlation_id",
+        lambda result, callback_data=callback_data: callback_data.complete(result)
+        )
+    callback_data.signal.wait()
+    return callback_data.auth_result
+
+
+def _convert_result(result, client_id):  # Mimic an on-the-wire response from AAD
+    error = result.get_error()
+    if error:
+        return _convert_error(error, client_id)
+    id_token_claims = json.loads(result.get_id_token()) if result.get_id_token() else {}
+    account = result.get_account()
+    assert account, "Account is expected to be always available"
+    ## Note: As of pymsalruntime 0.1.0, only wam_account_ids property is available
+    #account.get_account_property("wam_account_ids")
+    return_value = {k: v for k, v in {
+        "access_token": result.get_access_token(),
+        "expires_in": result.get_access_token_expiry_time(),
+        "id_token_claims": id_token_claims,
+        "client_info": account.get_client_info(),
+        "_account_id": account.get_account_id(),
+        }.items() if v}
+    granted_scopes = result.get_granted_scopes()  # New in pymsalruntime 0.3.x
+    if granted_scopes:
+        return_value["scope"] = " ".join(granted_scopes)  # Mimic the on-the-wire data format
+    return return_value
+
+
+def _get_new_correlation_id():
+    return str(uuid.uuid4())
+
+
+def _signin_silently(authority, client_id, scopes, correlation_id=None, **kwargs):
+    params = pymsalruntime.MSALRuntimeAuthParameters(client_id, authority)
+    params.set_requested_scopes(scopes)
+    callback_data = _CallbackData()
+    for k, v in kwargs.items():  # This can be used to support domain_hint, max_age, etc.
+        if v is not None:
+            params.set_additional_parameter(k, str(v))
+    pymsalruntime.signin_silently(
+        params,
+        correlation_id or _get_new_correlation_id(),
+        lambda result, callback_data=callback_data: callback_data.complete(result))
+    callback_data.signal.wait()
+    return _convert_result(callback_data.auth_result, client_id)
+
+
+def _signin_interactively(
+        authority, client_id, scopes,
+        window=None,
+        prompt=None,
+        login_hint=None,
+        claims=None,
+        correlation_id=None,
+        **kwargs):
+    params = pymsalruntime.MSALRuntimeAuthParameters(client_id, authority)
+    params.set_requested_scopes(scopes)
+    params.set_redirect_uri("placeholder")  # pymsalruntime 0.1 requires non-empty str,
```

Comment:
Because the underlying WAM behavior happens to be throw away any redirect_uri value the app passes it, I chose to literally use some irrelevant "placeholder" here to make a point.

If we pass in the app's redirect_uri here, it will give future maintainer an illusion that the underlying broker would use that redirect_uri value.

#Resolve

Reply:
Yes, WAM will throw away but it may not be the same as other broker like Linux. I think the API surface with the APP need to be maintain the same such that the parameters could be used in other broker instead of changing the API signature.

Reply:
> but it may not be the same as other broker like Linux.

When pymsalruntime starts to bridge the broker on Linux, our customer would NOT be affected by this line of code, because MSAL Python conservatively declares pymsalruntime dependency only when running on Windows.

This line would fail in OUR development device when we are porting this implementation to Linux. And we will refactor this at that time. That is kind of the TDD effect that we want.

> the API surface with the APP need to be maintain the same ... instead of changing the API signature.

None of this implementation detail would affect MSAL Python's API signature.

We can sync up on this tomorrow. :-)